spec/API_specification/array_api/array_object.py

```diff
@@ -473,37 +500,57 @@ def __gt__(self: array, other: Union[int, float, array], /) -> array:
            Element-wise results must equal the results returned by the equivalent element-wise function :func:`~array_api.greater`.
         """
 
-    def __index__(self: array, /) -> int:
+    def __index__(self: array, /) -> Union[int, bool]:
         """
-        Converts a zero-dimensional integer array to a Python ``int`` object.
+        Converts a zero-dimensional array to a Python object used in indexing.
 
         .. note::
            This method is called to implement `operator.index() <https://docs.python.org/3/reference/datamodel.html#object.__index__>`_. See also `PEP 357 <https://www.python.org/dev/peps/pep-0357/>`_.
 
         Parameters
         ----------
         self: array
-            zero-dimensional array instance. Must have an integer data type.
+            zero-dimensional array instance. Should have an integer or boolean data type. If ``self`` has a floating-point or complex data type, the function must raise a ``TypeError``.
 
         Returns
         -------
-        out: int
-            a Python ``int`` object representing the single element of the array instance.
+        out: Union[int, bool]
+            a Python object representing the single element of the array instance. If ``self`` has an integer data type, ``out`` must be a Python ``int`` object. If ``self`` has a boolean data type, ``out`` must be a Python ``bool`` object.
```

Comment:
I thought we didn't even specify behavior for indexing a single boolean. IMO we should just leave `__index__` as it is.

Reply:
Also `np.array(True).__index__()` raises an error and `np.bool_(True).__index__()` raises a deprecation error. In Python, `True.__index__()` returns `1`, but this is a mistake IMO because the whole point of `__index__()` is to only interpret actual integers as integers. In array indexing, booleans operate differently than integers. 

So I think we should definitely omit booleans from `__index__`, maybe even not allow optionally.

Reply:
It looks like the way it was written before does this ("self must have an integer data type"). So I would just revert the changes to `__index__`.

Reply:
Okay makes sense, reverted back to `__index__() -> int` behaviour.

Only thing I think we should do is leave boolean array behaviour unspecified, given Python does support `bool.__index__()` and a libraries might support that. `torch` currently supports `__index__()` on 0d bool tensors, returning a `int`.

```python
torch.as_tensor(True).__index__()
1
```

Reply:
I'd say that's a mistake on pytorch's part. `True` is a valid array index, and it doesn't mean the same thing as `1`. Single boolean indices are even mentioned explicitly in the spec.

Reply:
I am pretty sure there are torch issues open about it (at least one).  There could be an argument for allowing this, since Python does.  But it is really tricky, because then you would need a way to say that something is an array efficiently.

Long story short, Python True/False is weird, the rest of the world should not do this probably.

Reply:
Couldn't find a relevant issue (hard to search for heh), maybe @lezcano would know/has general thoughts?

But yeah I see it's weird behaviour, and we've got a nice opportunity here to forcibly disallow it. Just because Python does allow and so does pytorch (who's a strong candidate to adopt the spec on it's top-level namespace), I would lean towards leaving `<bool array>.__index__()` behaviour unspecified for now, and raise an issue on forcibly raising `TypeError` like the floating scenarios, given we did agree to emulate Python behaviour.

Reply:
Not sure it matters, but in case it interests someone, I think it is this one: https://github.com/pytorch/pytorch/issues/45125

Reply:
Sorry for the very late reply. This just fell through my notifications. Alas, I don't know enough as to say anything intelligent here, and even worse, I don't think that anyone currently working in PyTorch core would know what's going on here.

My best take is that, whoever implemented `__index__` went with Python's default behaviour in this case. Either that, or forgot to add a precondition.

Reply:
Limiting to `__index__` to only integers makes sense to me. We can revisit in a subsequent issue/thread if and when this topic arises again. For now, I'd consider the current language fine as is.